Dagger hilt

exRomeo opened the pull request.

exRomeo committed:
added city, country and Weather data classes

exRomeo committed:
added image assets and working on Weather Card and SearchBar Components

exRomeo committed:
added application theme colors in the corresponding files, working on search bar

exRomeo committed:
searchbar behaviour when receives results it expands with the results list
also added the searchbar Shape in application theme files
added a function to return dummy data inside countries class just for previews

exRomeo committed:
enhanced searchbar's drop down menu changed countries api to get more results and added an interceptor to retry failed http requests

exRomeo committed:
Added ui state sealed class with cases for idle, failure, success, no network, and loading to be used to reflect the current application state on the ui also added connection and error handling utilities and added loading screen and message screen to display different messages to the user such as there is no connection or an error occurred

exRomeo committed:
removed logs and fixed a bug where state would update immediately to success when the app launches and cleaned up extra unneeded parameters

exRomeo committed:
Enhanced weather results speed with asynchronous search tasks,
also centered Text in messages screen

exRomeo committed:
Added SlantShape as a composable function

exRomeo committed:
Updated SlantShape.kt,
enhanced the layout of WeatherCard

exRomeo committed:
updated SlantShape.kt
and used it in WeatherCard

exRomeo committed:
Restructured the code to prepare for MVI Architecture and created a state holder for SearchBar still under experiment

exRomeo committed:
Handled search states Idle, Loading Error and Success and represented each on the ui also added ui events to handle presenting all error messages to the user and show them on a snackbar
used UIAction to handle all actions made from ui the screen in the view model
todo: make a SnackbarWithAction to give the user the ability to retry failed actions

exRomeo committed:
added duration "long" to the snackbar and refactored some properties in home view model

exRomeo committed:
Applied Dagger 2 Dependency Injection

exRomeo committed:
Applied Dagger Hilt Dependency Injection

exRomeo committed:
updated search bar to display the current country's name in its place holder

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/remotesource/apiservice/apininja/ApiNinja.kt:
```diff
@@ -0,0 +1,33 @@
+package com.trianglz.weatherapp.data.remotesource.apiservice.apininja
```
In clean arch, the api service is an interface to invert the dependency of the remote source on the api service. 
So, the api service package should be a  folder outside the remote source folder. 
The layers are like this repo -> remote source -> api service -> actual retrofit 3rd party code.
But in the data package the api service will have a folder of it's own

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/repository/IRepository.kt:
```diff
@@ -0,0 +1,16 @@
+import com.trianglz.weatherapp.data.models.weather.Weather
+import com.trianglz.weatherapp.domain.utils.resource.Resource
+
+interface IRepository {
```
To achieve better isolation, separate ports (repositories) for each responsibility. We have a one way path for each feature.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/repository/Repository.kt:
```diff
@@ -0,0 +1,86 @@
+import kotlinx.coroutines.coroutineScope
+import javax.inject.Inject
+
+class Repository @Inject constructor(
```
In clean arch, this repo doesn't know the existence of an android API or network connectivity. So, checking for internet connection here isn't optimal. 
Although checking for the presence of a connection beforehand is preferred. Maybe in the remote source section.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/repository/Repository.kt:
```diff
@@ -0,0 +1,86 @@
+
+    override suspend fun getCountries(
+        countryName: String
+    ): Resource<List<Country>> = coroutineScope {
```
Avoid using custom monads for simple purposes. As already, the kotlin standard library provide one.

gemy411 replied:
Also, retrofit already returns an exception if no internet connection is established. 
And for fake connections (large wait time till timeout) retrofit already handles that. 
Unless the internet utility class actually knows a packet can be sent or received.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/models/weather/Weather.kt:
```diff
@@ -0,0 +1,49 @@
+    @field:SerializedName("sunset")
+    val sunset: Long,
+    var countryCode: String? = null,
+    var cityName: String? = null
```
having Var in a parse-able model usage in unclear.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/data/repository/Repository.kt:
```diff
@@ -0,0 +1,86 @@
+
+    private suspend fun getWeather(
+        city: City
+    ): Weather? {
```
Using the retrofit weather model in the repository is not preferred

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/domain/utils/connection/ConnectionUtility.kt:
```diff
@@ -0,0 +1,23 @@
+class ConnectionUtility @Inject constructor(application: Application) : IConnectionUtility {
+
+    private val connectivityManager = application.getSystemService(
+        ConnectivityManager::class.java
```
Regradless of the usage context of this utility class. `ConnectivityManager` can be injected using hilt.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/presentation/searchbarstate/SearchState.kt:
```diff
@@ -0,0 +1,42 @@
+import com.trianglz.weatherapp.presentation.ui.components.SearchBarStatus
+
+data class SearchState(
+    var placeHolder: String = "",
```
It's preferred in compose states that all the class properties are immutable. To avoid extra recompositions. 
Also, having everything immutable, will make you easily trace UI bugs. 
Recommended action: Do a little research how to wrap the composable state in model the clean way.

gemy411 commented on app/src/main/java/com/trianglz/weatherapp/presentation/searchbarstate/SearchState.kt:
```diff
@@ -0,0 +1,42 @@
+@Composable
+fun rememberSearchState(
+    initialState: SearchState = SearchState()
+) = remember(initialState) {
```
The key you pass to the remember function is the object that if changes will trigger the block to re-execute again. 
This code works only because you have `var`s in the model which make the composable engine mark this class as unstable and any mutation to the values will cause a cascade of re-compositions to ensure UI is updated properly.

exRomeo committed:
first attempt at Clean Architecture

- moved repository interface to domain package
- added usecases to adhere to separation of concern and separate use cases from repository class

exRomeo committed:
Edited an Exception message to better explain the reason for the Exception

removed unnecessary state flow from HomeViewModel

exRomeo committed:
Created Domain Models for country and weather separate from network DTOs and created mappers for them also replaced the DTOs with the Domain models in Domain and Presentation Layers

removed network connectivity check as its redundant
retrofit throws IOException when there is no internet connection

in SearchBarState changed all vars to vals

exRomeo committed:
Replaced Custom class Resource with Kotlin's Result Class

exRomeo committed:
Added navigation system and navigation drawer

exRomeo committed:
updated the app's theme so the background gradient continues behind the status bar

exRomeo committed:
Mass Refactoring
all data package modes renamed to have 'DataModel' in its name instead of 'Dto'

all Domain Models renamed to have 'DomainModel' in its name

renamed mappers according to the models new names

also renamed all interface implementations to have 'Impl' in its name and removed the 'I' from the interfaces names

also updated Compose Dependencies which required an sdkTarget upgrade which also required gradle upgrade
